pkg/skaffold/render/renderer/renderer.go

```diff
@@ -99,66 +94,37 @@ type SkaffoldRenderer struct {
+	// This is needed because kpt v1 only supports in-place config while users may not want to have their config be
+	// hydrated in place.
+	// Once Kptfile is initialized, its "pipeline" field will be updated in each skaffold render, and its "inventory"
+	// will keep the same to guarantee accurate `kpt live apply`.
+	_, endTrace := instrumentation.StartTrace(ctx, "Render_initKptfile")
 	if _, err := os.Stat(r.hydrationDir); os.IsNotExist(err) {
-		logrus.Debugf("creating render directory: %v", r.hydrationDir)
 		if err := os.MkdirAll(r.hydrationDir, os.ModePerm); err != nil {
-			return fmt.Errorf("creating render directory for hydration: %w", err)
+			endTrace(instrumentation.TraceEndError(fmt.Errorf("create hydration dir %v:%w", r.hydrationDir, err)))
+			return err
 		}
-	}
-	kptFilePath := filepath.Join(r.hydrationDir, kptfile.KptFileName)
-	if _, err := os.Stat(kptFilePath); os.IsNotExist(err) {
 		cmd := exec.CommandContext(ctx, "kpt", "pkg", "init", r.hydrationDir)
-		if _, err := util.RunCmdOut(cmd); err != nil {
-			return sErrors.NewError(err,
-				proto.ActionableErr{
-					Message: fmt.Sprintf("unable to initialize Kptfile in %v", r.hydrationDir),
-					ErrCode: proto.StatusCode_RENDER_KPTFILE_INIT_ERR,
-					Suggestions: []*proto.Suggestion{
-						{
-							SuggestionCode: proto.SuggestionCode_KPTFILE_MANUAL_INIT,
-							Action:         fmt.Sprintf("please manually run `kpt pkg init %v`", r.hydrationDir),
-						},
-					},
-				})
+		if err := util.RunCmd(cmd); err != nil {
+			endTrace(instrumentation.TraceEndError(fmt.Errorf("`kpt pkg init %v`:%w", r.hydrationDir, err)))
+			return err
 		}
 	}
-	return nil
-}
-
-func (r *SkaffoldRenderer) Render(ctx context.Context, out io.Writer, builds []graph.Artifact) error {
-	if err := r.prepareHydrationDir(ctx); err != nil {
-		return err
-	}
-
-	manifests, err := r.Generate(ctx)
-	if err != nil {
-		return err
-	}
-	manifests, err = manifests.ReplaceImages(ctx, builds)
+	endTrace()
+	_, endTrace = instrumentation.StartTrace(ctx, "Render_readKptfile")
+	file, err := os.Open(kptfilePath)
 	if err != nil {
+		endTrace(instrumentation.TraceEndError(fmt.Errorf("read Kptfile from %v: %w",
+			filepath.Dir(kptfilePath), err)))
 		return err
 	}
-	manifests.SetLabels(r.labels)
-
-	// cache the dry manifests to the temp directory. manifests.yaml will be truncated if already exists.
-	dryConfigPath := filepath.Join(r.hydrationDir, dryFileName)
-	if err := manifest.Write(manifests.String(), dryConfigPath, out); err != nil {
-		return err
-	}
-
-	// Read the existing Kptfile content. Kptfile is guaranteed to be exist in prepareHydrationDir.
-	kptFilePath := filepath.Join(r.hydrationDir, kptfile.KptFileName)
-	file, err := os.Open(kptFilePath)
-	if err != nil {
-		return err
-	}
-	defer file.Close()
-	kfConfig := &kptfile.KptFile{}
+	defer func() { _ = file.Close() }()
 	if err := yaml.NewDecoder(file).Decode(&kfConfig); err != nil {
```

Comment:
if KptFile schema changes in future how do we keep this in sync? It seems to only parse a specific version.

Reply:
Yes, that's the plan (only parse a specific version). After kpt is GA-ed, it would then support backward compatibility for most cases so I won't worry too much about this. And we are also planning to have kpt as a builtin go bindata (may not in the render v2 alpha release)

To answer your question more generally. kpt shall provide a go package/lib to allow users to get the Kptfile. Kpt team is aware of this usage but this go package needs much more work and is not in their v1 release plan.

Reply:
my concern is not about backward compatibility for this version of the config with a newer kpt binary; rather that a user comes with a newer version of `Kptfile` which this code won't be able to parse. That'd be okay if this Kptfile is always skaffold generated. Or if the config structs are never modified (only new fields get added) across versions.

Reply:
The scenario you mentioned is not covered in this case. Manual change may be needed if skaffold plans to support a newer kpt version (note, kpt schema shall not be updated that frequently as skaffold).